File: implementations/elixir/lib/rcl_convert.ex
defmodule RCL.Convert do
  def to_yaml(m), do: yaml_map(m, 0)

  def to_toml(m) do
    lines = toml_table(m, [])
    if lines == [], do: "", else: Enum.join(lines, "\n") <> "\n"
  end

  def to_hcl(m) do
    if m == %{}, do: "{}\n", else: Enum.map_join(entries(m), "", fn {k, v} -> hcl_key(k) <> " = " <> hcl_val(v, 0) <> "\n" end)
  end

  defp yaml_map(m, n) do
    if m == %{} do
      String.duplicate(" ", n) <> "{}\n"
    else
      Enum.map_join(entries(m), "", fn {k, v} ->
        i = String.duplicate(" ", n)
        if is_map(v), do: i <> yaml_key(k) <> ":\n" <> yaml_map(v, n + 2), else: i <> yaml_key(k) <> ": " <> scalar(v) <> "\n"
      end)
    end
  end

  defp toml_table(m, path) do
    header = if path == [], do: [], else: ["[" <> Enum.map_join(path, ".", &toml_key/1) <> "]"]
    {maps, vals} = Enum.reduce(entries(m), {[], []}, fn {k, v}, {ms, vs} -> if is_map(v), do: {[{k, v} | ms], vs}, else: {ms, [toml_key(k) <> " = " <> scalar(v) | vs]} end)
    base = header ++ Enum.reverse(vals)
    Enum.reverse(maps)
    |> Enum.reduce(base, fn {k, v}, acc ->
      t = toml_table(v, path ++ [k])
      if acc == [], do: t, else: acc ++ [""] ++ t
    end)
  end

  defp hcl_val(v, n) do
    cond do
      is_map(v) and map_size(v) == 0 -> "{}"
      is_map(v) ->
        i = String.duplicate(" ", n)
        "{\n" <> Enum.map_join(entries(v), "", fn {k, x} -> i <> "  " <> hcl_key(k) <> " = " <> hcl_val(x, n + 2) <> "\n" end) <> i <> "}"
      true -> scalar(v)
    end
  end

  defp scalar(v) when is_binary(v), do: "\"" <> esc(v) <> "\""
  defp scalar(v) when is_boolean(v), do: if(v, do: "true", else: "false")
  defp scalar(v) when is_list(v), do: "[" <> Enum.map_join(v, ", ", &scalar/1) <> "]"
  defp scalar(v), do: to_string(v)

  defp entries(m), do: Enum.sort_by(m, fn {k, _} -> k end)
  defp bare?(k), do: Regex.match?(~r/^[A-Za-z_][A-Za-z0-9_]*$/, k)
  defp toml_key(k), do: if(bare?(k), do: k, else: "\"" <> esc(k) <> "\"")
  defp yaml_key(k), do: if(bare?(k), do: k, else: "\"" <> esc(k) <> "\"")
  defp hcl_key(k), do: if(bare?(k), do: k, else: "\"" <> esc(k) <> "\"")
  defp esc(s), do: s |> String.replace("\\", "\\\\") |> String.replace("\"", "\\\"") |> String.replace("\n", "\\n") |> String.replace("\t", "\\t")
end

File: implementations/php/src/RCLConvert.php
<?php

final class RCLConvert {
  public static function yaml(array $obj): string {
    return self::yamlMap($obj, 0);
  }
  public static function toml(array $obj): string {
    $lines = self::tomlTable($obj, []);
    return $lines === [] ? '' : implode("\n", $lines) . "\n";
  }
  public static function hcl(array $obj): string {
    if ($obj === []) return "{}\n";
    $s = '';
    foreach (self::entries($obj) as [$k, $v]) $s .= self::hclKey($k) . ' = ' . self::hclVal($v, 0) . "\n";
    return $s;
  }
  private static function yamlMap(array $m, int $n): string {
    if ($m === []) return str_repeat(' ', $n) . "{}\n";
    $s = '';
    foreach (self::entries($m) as [$k, $v]) {
      $i = str_repeat(' ', $n);
      if (is_array($v) && self::isAssoc($v)) $s .= $i . self::yamlKey($k) . ":\n" . self::yamlMap($v, $n + 2);
      else $s .= $i . self::yamlKey($k) . ': ' . self::yamlScalar($v) . "\n";
    }
    return $s;
  }
  private static function tomlTable(array $m, array $path): array {
    $lines = [];
    if ($path !== []) $lines[] = '[' . implode('.', array_map(fn($k) => self::tomlKey($k), $path)) . ']';
    $children = [];
    foreach (self::entries($m) as [$k, $v]) {
      if (is_array($v) && self::isAssoc($v)) $children[] = [$k, $v];
      else $lines[] = self::tomlKey($k) . ' = ' . self::tomlScalar($v);
    }
    foreach ($children as [$k, $v]) {
      $child = self::tomlTable($v, [...$path, $k]);
      if ($lines !== [] && end($lines) !== '') $lines[] = '';
      $lines = [...$lines, ...$child];
    }
    return $lines;
  }
  private static function hclVal(mixed $v, int $n): string {
    if (is_array($v) && self::isAssoc($v)) {
      if ($v === []) return '{}';
      $i = str_repeat(' ', $n);
      $s = "{\n";
      foreach (self::entries($v) as [$k, $x]) $s .= $i . '  ' . self::hclKey($k) . ' = ' . self::hclVal($x, $n + 2) . "\n";
      return $s . $i . '}';
    }
    return self::tomlScalar($v);
  }
  private static function tomlScalar(mixed $v): string {
    if (is_string($v)) return '"' . self::esc($v) . '"';
    if (is_bool($v)) return $v ? 'true' : 'false';
    if (is_array($v)) return '[' . implode(', ', array_map(fn($x) => self::tomlScalar($x), $v)) . ']';
    return (string)$v;
  }
  private static function yamlScalar(mixed $v): string {
    if (is_array($v)) return '[' . implode(', ', array_map(fn($x) => self::yamlScalar($x), $v)) . ']';
    return self::tomlScalar($v);
  }
  private static function esc(string $s): string {
    return str_replace(["\\", '"', "\n", "\t"], ["\\\\", '\\"', '\\n', '\\t'], $s);
  }
  private static function isAssoc(array $a): bool { return array_keys($a) !== range(0, count($a) - 1); }
  private static function entries(array $m): array { ksort($m); $out = []; foreach ($m as $k => $v) $out[] = [(string)$k, $v]; return $out; }
  private static function bare(string $k): bool { return preg_match('/^[A-Za-z_][A-Za-z0-9_]*$/', $k) === 1; }
  private static function tomlKey(string $k): string { return self::bare($k) ? $k : '"' . self::esc($k) . '"'; }
  private static function yamlKey(string $k): string { return self::bare($k) ? $k : '"' . self::esc($k) . '"'; }
  private static function hclKey(string $k): string { return self::bare($k) ? $k : '"' . self::esc($k) . '"'; }
}
